### PR TITLE
Compute MACs for full input/output tensor

### LINKED ISSUE #32: Torchinfo does not take sequence length for LSTM models into account for computing MACs
First of all, great work @TylerYep - torchinfo looks very nice :)

I have a question regarding the MAC computation for LSTM networks. Somehow it seems that the sequence length is not taken into account.

Here is the example from the `README.md` with two different sequence lengths:

```python
class LSTMNet(nn.Module):
    """ Batch-first LSTM model. """
    def __init__(self, vocab_size=20, embed_dim=300, hidden_dim=512, num_layers=2):
        super().__init__()
        self.hidden_dim = hidden_dim
        self.embedding = nn.Embedding(vocab_size, embed_dim)
        self.encoder = nn.LSTM(embed_dim, hidden_dim, num_layers=num_layers, batch_first=True)
        self.decoder = nn.Linear(hidden_dim, vocab_size)

    def forward(self, x):
        embed = self.embedding(x)
        out, hidden = self.encoder(embed)
        out = self.decoder(out)
        out = out.view(-1, out.size(2))
        return out, hidden

seq_length = 100
summary(
    LSTMNet(),
    (1, seq_length),
    dtypes=[torch.long],
    verbose=2,
    col_width=16,
    col_names=["kernel_size", "output_size", "num_params", "mult_adds"],
)

seq_length = 10
summary(
    LSTMNet(),
    (1, seq_length),
    dtypes=[torch.long],
    verbose=2,
    col_width=16,
    col_names=["kernel_size", "output_size", "num_params", "mult_adds"],
)
```

This gives the following output:
```
========================================================================================================
Layer (type:depth-idx)                   Kernel Shape     Output Shape     Param #          Mult-Adds
========================================================================================================
\u251c\u2500Embedding: 1-1                         [300, 20]        [1, 100, 300]    6,000            6,000
\u251c\u2500LSTM: 1-2                              --               [1, 100, 512]    3,768,320        3,760,128
|    \u2514\u2500weight_ih_l0                      [2048, 300]
|    \u2514\u2500weight_hh_l0                      [2048, 512]
|    \u2514\u2500weight_ih_l1                      [2048, 512]
|    \u2514\u2500weight_hh_l1                      [2048, 512]
\u251c\u2500Linear: 1-3                            [512, 20]        [1, 100, 20]     10,260           10,240
========================================================================================================
Total params: 3,784,580
Trainable params: 3,784,580
Non-trainable params: 0
Total mult-adds (M): 3.78
========================================================================================================
Input size (MB): 0.00
Forward/backward pass size (MB): 0.67
Params size (MB): 15.14
Estimated Total Size (MB): 15.80
========================================================================================================

========================================================================================================
Layer (type:depth-idx)                   Kernel Shape     Output Shape     Param #          Mult-Adds
========================================================================================================
\u251c\u2500Embedding: 1-1                         [300, 20]        [1, 10, 300]     6,000            6,000
\u251c\u2500LSTM: 1-2                              --               [1, 10, 512]     3,768,320        3,760,128
|    \u2514\u2500weight_ih_l0                      [2048, 300]
|    \u2514\u2500weight_hh_l0                      [2048, 512]
|    \u2514\u2500weight_ih_l1                      [2048, 512]
|    \u2514\u2500weight_hh_l1                      [2048, 512]
\u251c\u2500Linear: 1-3                            [512, 20]        [1, 10, 20]      10,260           10,240
========================================================================================================
Total params: 3,784,580
Trainable params: 3,784,580
Non-trainable params: 0
Total mult-adds (M): 3.78
========================================================================================================
Input size (MB): 0.00
Forward/backward pass size (MB): 0.07
Params size (MB): 15.14
Estimated Total Size (MB): 15.20
========================================================================================================
```

Both times, the `Total mult-adds (M)` is the same, although the sequence length is different. I think the problem is that torchinfo ignores the `seq_length`, i.e., does not take into account for example the nn.Linear is used `seq_length` times. Is this behavior expected?

### PR DESCRIPTION
**Pull request by TE-StefanUhlich**
This changes the value that is returned by `summary`. Up to now, this value was assuming a batch-size of 1 and, thus, ignored the batch size in the MAC computations. However, this does not work with recurrent NNs as these, e.g., share fully connected layers over many timesteps.

With this change, the correct numbers are printed:
```
seq_length = 100:
========================================================================================================
Layer (type:depth-idx)                   Kernel Shape     Output Shape     Param #          Mult-Adds
========================================================================================================
\u251c\u2500Embedding: 1-1                         [300, 20]        [1, 100, 300]    6,000            6,000
\u251c\u2500LSTM: 1-2                              --               [1, 100, 512]    3,768,320        376,012,800
|    \u2514\u2500weight_ih_l0                      [2048, 300]
|    \u2514\u2500weight_hh_l0                      [2048, 512]
|    \u2514\u2500weight_ih_l1                      [2048, 512]
|    \u2514\u2500weight_hh_l1                      [2048, 512]
\u251c\u2500Linear: 1-3                            [512, 20]        [1, 100, 20]     10,260           10,240
========================================================================================================
Total params: 3,784,580
Trainable params: 3,784,580
Non-trainable params: 0
Total mult-adds (M): 376.03
========================================================================================================
Input size (MB): 0.00
Forward/backward pass size (MB): 0.67
Params size (MB): 15.14
Estimated Total Size (MB): 15.80
========================================================================================================
```

```
seq_length=10:
========================================================================================================
Layer (type:depth-idx)                   Kernel Shape     Output Shape     Param #          Mult-Adds
========================================================================================================
\u251c\u2500Embedding: 1-1                         [300, 20]        [1, 10, 300]     6,000            6,000
\u251c\u2500LSTM: 1-2                              --               [1, 10, 512]     3,768,320        37,601,280
|    \u2514\u2500weight_ih_l0                      [2048, 300]
|    \u2514\u2500weight_hh_l0                      [2048, 512]
|    \u2514\u2500weight_ih_l1                      [2048, 512]
|    \u2514\u2500weight_hh_l1                      [2048, 512]
\u251c\u2500Linear: 1-3                            [512, 20]        [1, 10, 20]      10,260           10,240
========================================================================================================
Total params: 3,784,580
Trainable params: 3,784,580
Non-trainable params: 0
Total mult-adds (M): 37.62
========================================================================================================
Input size (MB): 0.00
Forward/backward pass size (MB): 0.07
Params size (MB): 15.14
Estimated Total Size (MB): 15.20
========================================================================================================
```

Fixes #32

### CODE REVIEW

**Comment by TE-StefanUhlich** [comment]
@TylerYep Shouldn't this [line](https://github.com/TylerYep/torchinfo/blob/main/setup.cfg#L3) be reading `version = 1.4.4`?

**Comment by TE-StefanUhlich** [comment]
I just noticed the failing tests: This is due to the change in definition of Multiply-Adds.
@TylerYep Is there an easy way to update the test files?

**Comment by TylerYep** [comment]
We renamed the project and add a result we restarted the version numbering.

The tests are available to update as well.

**Comment by TE-StefanUhlich** [comment]
> The tests are available to update as well.

Yes, I have seen them in `tests` - I was just wondering if there is an easy way to recreate the output files in [`test_output`](https://github.com/TylerYep/torchinfo/tree/main/tests/test_output). How did you create them?

**Comment by TylerYep** [comment]
I ran the test (in pytest, using -s and -vv), and copy + pasted the output into a new output file.

**Comment by TE-StefanUhlich** [comment]
The output tests are fixed - thanks for your help.

But there is one test that still fails:

```bash
$ CUDA_VISIBLE_DEVICES= pytest
...
torchinfo/torchinfo.py:349: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = ReturnDictLayer: 1-1

    def calculate_macs(self) -> None:
        """
        Set MACs using the module's parameters and layer's output size, which is
        used for computing number of operations for Conv layers.
    
        Please note: Returned MACs is the number of MACs for the full
                     tensor, i.e., taking the batch-dimension into account.
        """
        for name, param in self.module.named_parameters():
            if name == "weight":
                # ignore C when calculating Mult-Adds in ConvNd
                if "Conv" in self.class_name:
                    self.macs += int(param.nelement() *
                                     prod(self.output_size[:1] + self.output_size[2:]))
                else:
                    self.macs += self.output_size[0] * param.nelement()
            # RNN modules have inner weights such as weight_ih_l0
            elif "weight" in name:
>               self.macs += prod(self.output_size[:2]) * param.nelement()

torchinfo/layer_info.py:134: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

num_list = [[1, 10]]

    def prod(num_list: Union[Iterable[Any], torch.Size]) -> int:
        result = 1
        for num in num_list:
            result *= num
>       return abs(result)
E       TypeError: bad operand type for abs(): 'list'
...
```

It seems that `self.output_size[:2]` is a list of lists - probably as it returns several outputs, right? How should we deal with this case?

**Comment by TylerYep** [comment]
We could probably modify the prod() function to flatten any nested sequences passed in.

Also, note that we need to keep the batch_dim of 2 in order for BatchNorm tests to pass. We don't want to include 2 in the calculations, so you should be able to use batch_dim  parameter to determine whether the user has given a batch dimension in order to exclude it accordingly.

**Comment by TE-StefanUhlich** [comment]
I fixed `prod` as discussed such that it now also can deal with list of lists. Furthermore, I noticed that some tests were not working if a GPU is available (this is due to `summary` using `device="cuda"` if a GPU is available (see [here](https://github.com/TylerYep/torchinfo/blob/main/torchinfo/torchinfo.py#L141)).


>Also, note that we need to keep the batch_dim of 2 in order for BatchNorm tests to pass. We don't want to include 2 in the calculations, so you should be able to use batch_dim parameter to determine whether the user has given a batch dimension in order to exclude it accordingly.

Can you please elaborate more on this issue? I don't see a problem with a batch-size of one as we operate the network in `eval`-mode. (even if we would be in training mode should be fine).

**Comment by TylerYep** [comment]
Thanks for the device="cpu" fix.

Batch size of 2 was done to address this problem: https://discuss.pytorch.org/t/batch-norm-training-when-batch-size-1/102979/9. This was a feature I adopted from some of the original projects, so not 100% certain whether this is needed.

**Comment by TE-StefanUhlich** [comment]
ok, I see. For `torchinfo`, I think there is not this problem as the network is turned into `eval`-mode [here](https://github.com/TylerYep/torchinfo/blob/main/torchinfo/torchinfo.py#L157), right?

**Comment by TylerYep** [comment]
I looked into the failure case and you are correct - this isn't an issue because we run the model in eval() mode.
```python
class BatchNormModel(nn.Module):
    def __init__(self) -> None:
        super().__init__()
        self.dense1_bn = nn.BatchNorm1d(3)

    def forward(self, x):
        return self.dense1_bn(x)

summary(BatchNormModel(), input_size=(3, 1), batch_dim=0)
# ValueError: Expected more than 1 value per channel when training, got input size torch.Size([1, 3, 1])
```

File: torchinfo/layer_info.py [hunk 1]
```diff
@@ -163,5 +167,8 @@ def num_params_to_str(self, reached_max_depth: bool = False) -> str:
 def prod(num_list: Union[Iterable[Any], torch.Size]) -> int:
     result = 1
     for num in num_list:
-        result *= num
-    return abs(result)
+        try:
```
Comment (TylerYep) [inline torchinfo/layer_info.py:170]: Preferred that we use a base case rather than catching an exception

**Review by TylerYep — APPROVED** [review]
Looks good to me, thank you for making this change!

**Pull request merged**